crates/energy/src/graph.rs

```diff
@@ -0,0 +1,242 @@
+                    update_graph.in_set(GraphSystemSet::UpdateGraph).after(GraphSystemSet::UpdateNearby),
+                    debug_lines.after(GraphSystemSet::UpdateGraph),
+                )
+                    .run_if(in_state(GameState::Playing)),
+            );
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+    UpdateGraph,
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy receiver component is used to mark an entity as an energy receiver.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyReceiver;
+
+/// The energy producer component is used to mark an entity as an energy producer.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyProducer;
+
+/// The nearby component is used to store the nearby entities of an entity.
+#[derive(Debug, Clone)]
+pub enum Nearby {
+    Receiver(SmallVec<[Entity; 40]>),
+    Producer(SmallVec<[Entity; 40]>),
+}
+
+impl Nearby {
+    fn into_inner(self) -> SmallVec<[Entity; 40]> {
+        match self {
+            Nearby::Receiver(inner) => inner,
+            Nearby::Producer(inner) => inner,
+        }
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits(SmallVec<[Nearby; 2]>, Option<Vec2>);
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Active>>,
+) {
+    for entity in newly_spawned_units.iter() {
+        commands
+            .entity(entity)
+            .insert((EnergyReceiver, NearbyUnits::default()));
+    }
+}
+
+fn update_nearby_recv(
+    spacial_index_producer: SpatialQuery<Entity, With<EnergyProducer>>,
+    spacial_index_receiver: SpatialQuery<Entity, With<EnergyReceiver>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform)>,
+) {
+    let time = Instant::now();
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            if let Some(last_pos) = nearby_units.1 {
+                if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {
+                    return;
+                }
+            }
+            nearby_units.1 = Some(transform.translation.to_flat());
+
+            let aabb = &Aabb::new(
+                Point::from(*(transform.translation - Vec3::splat(MAX_DISTANCE)).as_ref()),
```

Comment:
`*(...).as_ref()` why is this needed?

Reply:
It is a vec3 `.as_ref()` turns it into a `&[f64;3]` and the method takes a [f64;3].

Reply:
I think there is `From<Vec3>` implementation for `Point`. https://docs.rs/nalgebra/0.32.3/nalgebra/geometry/type.Point3.html#impl-From%3CVec3%3E-for-Point3%3Cf32%3E-1

Reply:
You are correct, I rely on my IDE quite heavily for indexing/finding methods and it has failed in this case.